pkg/querier/querier.go

```diff
@@ -505,31 +516,68 @@ func (s *sliceSeriesSet) Warnings() storage.Warnings {
-	now := model.Now()
-	startTime := model.Time(startMs)
-	endTime := model.Time(endMs)
-
-	endTime = clampTime(ctx, endTime, maxQueryIntoFuture, now.Add(maxQueryIntoFuture), false, "end", "max query into future", logger)
+func validateQueryTimeRange(userID string, startMs, endMs int64, limits *validation.Overrides, maxQueryIntoFuture time.Duration, logger log.Logger) (int64, int64, error) {
+	now := time.Now().UnixMilli()
+	endMs = clampTime(logger, endMs, now, maxQueryIntoFuture, "max query into future", true)
 
 	maxQueryLookback := limits.MaxQueryLookback(userID)
-	startTime = clampTime(ctx, startTime, maxQueryLookback, now.Add(-maxQueryLookback), true, "start", "max query lookback", logger)
+	startMs = clampTime(logger, startMs, now, -maxQueryLookback, "max query lookback", false)
 
-	if endTime.Before(startTime) {
+	if endMs < startMs {
 		return 0, 0, errEmptyTimeRange
 	}
 
-	return int64(startTime), int64(endTime), nil
+	return startMs, endMs, nil
 }
 
-// Ensure a time is within bounds, and log in traces to ease debugging.
-func clampTime(ctx context.Context, t model.Time, limit time.Duration, clamp model.Time, before bool, kind, name string, logger log.Logger) model.Time {
-	if limit > 0 && ((before && t.Before(clamp)) || (!before && t.After(clamp))) {
-		level.Debug(spanlogger.FromContext(ctx, logger)).Log(
-			"msg", "the "+kind+" time of the query has been manipulated because of the '"+name+"' setting",
-			"original", util.FormatTimeModel(t),
-			"updated", util.FormatTimeModel(clamp))
-		t = clamp
+// clampTime allows for time-limiting query minT and maxT based on configured limits.
+//
+// origT is the original timestamp to be clamped based on other supplied parameters.
+//
+// refT is the reference time from which to apply limitDelta; generally now() or query maxT because
+// when we truncate on the basis of overall query length, we leave maxT as-is and bring minT forward.
+//
+// clampDelta should be negative if the limit is looking back in time from the reference time;
+// this most common, as we mostly truncate a minT forward or a maxT backward by applying
+// a limit looking backwards from a reference time of now() or query maxT, as in:
+//   - max-query-lookback: (refTime: now)
+//   - query-ingesters-within (refTime: now)
+//   - query-store-after (refTime: now)
+//   - max-labels-query-length (refTime: maxT)
+//
+// An example exception is max-query-into-future, where the maxT is truncated by applying
+// the limit max-query-into-future forward in time from the reference time of now().
+func clampTime(
+	ll log.Logger, origT int64, refT int64, limitDelta time.Duration, limitName string, clampMax bool,
+) int64 {
+	if limitDelta == 0 {
+		// limits equal to 0 are considered to not be enabled
+		return origT
+	}
+	allowedT := refT + limitDelta.Milliseconds()
+
+	clampedT := origT
+	if allowedT < origT && clampMax {
+		// moving a max time window backwards
+		clampedT = math.Min(origT, allowedT)
```

Comment:
Can't we just set `clampedT = allowedT` here? Don't we know that `allowedT` is smaller than `origT` in this branch? (Similar comment applies for `else` branch below.)